src/java/org/apache/cassandra/db/ColumnFamilyStore.java

```diff
@@ -891,14 +897,14 @@ private void logFlush()
      * @return a Future yielding the commit log position that can be guaranteed to have been successfully written
      *         to sstables for this table once the future completes
      */
-    public ListenableFuture<CommitLogPosition> forceFlush()
+    public ListenableFuture<CommitLogPosition> forceFlush(FlushReason reason)
     {
         synchronized (data)
         {
             Memtable current = data.getView().getCurrentMemtable();
             for (ColumnFamilyStore cfs : concatWithIndexes())
                 if (!cfs.data.getView().getCurrentMemtable().isClean())
-                    return switchMemtableIfCurrent(current);
+                    return flushMemtable(current, reason);
```

Comment:
I was always baffled by the flushing code.

Why can't it be split into two parts - flush memtable if necessary and `waitForFlushes()`?
I.e. - is returning `ListenableFuture<CommitLogPosition>` from memtable switching instead of one from `waitForFlushes` an "optimization", real optimization, or is it important because of correctness?

Currently we `waitForFlushes` in 5 different places on different levels of abstraction and also get the commit log position from memtable switching directly.

Reply:
In addition to not having to insert another task in the queue, the one returned by the flush is precise in the sense that it provides the exact position of the flush.

OTOH I don't know how critical that is.